File: go/api/Vpn.go
package api

import (
	"github.com/lvcdy/huawei-lte-api-go/enums"
	"github.com/lvcdy/huawei-lte-api-go/session"
)

// Vpn 对应 Vpn.py。
type Vpn struct {
	*session.ApiGroup
}

// NewVpn 创建 Vpn API 分组。
func NewVpn(s *session.Session) *Vpn {
	return &Vpn{ApiGroup: session.NewApiGroup(s)}
}

// FeatureSwitch 对应 vpn/feature-switch。
func (v *Vpn) FeatureSwitch() (map[string]interface{}, error) {
	return v.S.Get("vpn/feature-switch", nil, "api")
}

// BrList 对应 vpn/br_list。
func (v *Vpn) BrList() (map[string]interface{}, error) {
	return v.S.Get("vpn/br_list", nil, "api")
}

// IpsecSettings 对应 vpn/ipsec_settings。
func (v *Vpn) IpsecSettings() (map[string]interface{}, error) {
	return v.S.Get("vpn/ipsec_settings", nil, "api")
}

// L2tpSettings 对应 vpn/l2tp_settings。
func (v *Vpn) L2tpSettings() (map[string]interface{}, error) {
	return v.S.Get("vpn/l2tp_settings", nil, "api")
}

// PptpSettings 对应 vpn/pptp_settings。
func (v *Vpn) PptpSettings() (map[string]interface{}, error) {
	return v.S.Get("vpn/pptp_settings", nil, "api")
}

// ToggleStatus 启用/禁用指定类型的 VPN。
// 对应 vpn/{pptp|l2tp}_settings (post_set, is_encrypted)。
func (v *Vpn) ToggleStatus(enable bool, vpnType enums.VPNType) (interface{}, error) {
	enableStr := "0"
	if enable {
		enableStr = "1"
	}
	return v.S.PostSet("vpn/"+string(vpnType)+"_settings", map[string]interface{}{
		"enable": enableStr,
	}, false, "api", true, false)
}

// Status 对应 vpn/status。
func (v *Vpn) Status() (map[string]interface{}, error) {
	return v.S.Get("vpn/status", nil, "api")
}

File: src/api/vpn.rs
//! Vpn API group (`api/Vpn.py`).

use serde_json::Value;

use crate::connection::Connection;
use crate::errors::Result;
use crate::tools::map_of;
use crate::xml::XmlMap;

use super::{get_value, post_set};

/// Vpn API group.
pub struct Vpn<'a> {
    conn: &'a Connection,
}

impl<'a> Vpn<'a> {
    /// Build a group bound to `conn`.
    pub fn new(conn: &'a Connection) -> Self {
        Vpn { conn }
    }

    /// `vpn/feature-switch`. VPN feature switch status.
    pub fn feature_switch(&self) -> Result<Value> {
        get_value(self.conn, "api/vpn/feature-switch")
    }

    /// `vpn/br_list`. VPN bridge list.
    pub fn br_list(&self) -> Result<Value> {
        get_value(self.conn, "api/vpn/br_list")
    }

    /// `vpn/ipsec_settings`. IPSec settings.
    pub fn ipsec_settings(&self) -> Result<Value> {
        get_value(self.conn, "api/vpn/ipsec_settings")
    }

    /// `vpn/l2tp_settings`. L2TP settings.
    pub fn l2tp_settings(&self) -> Result<Value> {
        get_value(self.conn, "api/vpn/l2tp_settings")
    }

    /// `vpn/pptp_settings`. PPTP settings.
    pub fn pptp_settings(&self) -> Result<Value> {
        get_value(self.conn, "api/vpn/pptp_settings")
    }

    /// Enable/disable a VPN type.
    ///
    /// * `enable` — `true` to enable, `false` to disable.
    /// * `vpn_type` — e.g. `"pptp"` or `"l2tp"` (mirrors the Python `VPNType.value`).
    pub fn toggle_status(&self, enable: bool, vpn_type: &str) -> Result<String> {
        let body: XmlMap = map_of([(
            "enable",
            if enable {
                "1".to_string()
            } else {
                "0".to_string()
            },
        )]);
        post_set(self.conn, &format!("api/vpn/{vpn_type}_settings"), &body)
    }

    /// `vpn/status`. VPN status (reverse engineered).
    pub fn status(&self) -> Result<Value> {
        get_value(self.conn, "api/vpn/status")
    }
}
